Adds branding customisation

m1kola opened the pull request:
~Also splits removebootstrap install phase into more steps~

m1kola commented on pkg/install/3-setuprouter.go:
```diff
@@ -0,0 +1,66 @@
+			return err
+		}
+
+		svc, err := cli.CoreV1().Services("openshift-ingress").Get("router-default", metav1.GetOptions{})
```
Needs changing. This will likely fail because it now happens before the step where we wait for clusterversion.

jim-minter commented on pkg/install/4-addcustomisations.go:
```diff
@@ -0,0 +1,95 @@
+		if err != nil {
+			return err
+		}
+	}
```
is there something here that we should wait for?  e.g. can we wait for an indication that the console has reloaded with the new branding before calling the cluster creation complete?

m1kola replied:
@jim-minter good question! Waiting for the console to reaload with ARO branding will add to the cluster creation time. 

Given that console should be accessible during the reload, is it critical if user can see the standard branding for a minute or two?

jim-minter replied:
please start by adding in the wait - let's see where we go from there.

jim-minter commented on pkg/install/3-setuprouter.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	installConfig := g[reflect.TypeOf(&installconfig.InstallConfig{})].(*installconfig.InstallConfig)
+	kubeadminPassword := g[reflect.TypeOf(&password.KubeadminPassword{})].(*password.KubeadminPassword)
```
could you pull this to the top to conform with the other functions?

m1kola commented:
@jim-minter I addressed the feedback. Please take another look. Now we wait on the console to reaload with the branding (adds ~1.5 minutes to the installation time).

jim-minter commented on Gopkg.lock:
```diff
@@ -832,6 +836,14 @@
   revision = "2336783d460353458477aa9e181eca87e4db5d27"
 
+[[projects]]
+  branch = "master"
```
pin this to release-4.3 in Gopkg.toml and revendor please

jim-minter commented on pkg/install/3-setuprouter.go:
```diff
@@ -0,0 +1,72 @@
+		return err
+	}
+
+	i.doc, err = i.db.PatchWithLease(ctx, i.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
```
Opened issue https://github.com/Azure/ARO-RP/issues/103

jim-minter commented on pkg/install/4-addcustomisations.go:
```diff
@@ -0,0 +1,116 @@
+		defer cancel()
+		err = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
+			operatorConfig, err := cli.OperatorV1().Consoles().Get(consoleapi.ConfigResourceName, metav1.GetOptions{})
+			if err == nil && operatorConfig.Status.ObservedGeneration > configGeneration {
```
I'd expect to see you test operatorConfig.Status.ObservedGeneration == operatorConfig.Generation here - is there a particular reason why not to?

jim-minter commented on pkg/install/4-addcustomisations.go:
```diff
@@ -0,0 +1,116 @@
+		}
+	}
+
+	{
```
idle thought (not requesting a change in this PR): I wonder if we will be able to pull this above waiting for the clusterversion in future.

jim-minter commented on pkg/install/4-addcustomisations.go:
```diff
@@ -0,0 +1,116 @@
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+func (i *Installer) addCustomisations(ctx context.Context) error {
```
if we're going to break these functions down as a prelude to unit tests, we can divide this one into two, I think?

jim-minter commented on pkg/api/openshiftcluster.go:
```diff
@@ -204,4 +204,6 @@ const (
 	InstallPhaseDeployStorage InstallPhase = iota
 	InstallPhaseDeployResources
 	InstallPhaseRemoveBootstrap
+	InstallPhaseSetupRouter
```
I am worried about adding more phases because I think they could easily play badly with RP upgrades.  If we have installs running in flight as we upgrade, and we add/remove phases, things could get pretty weird.

I wonder whether we should be trying to reduce to two phases - "everything before we delete the bootstrap", and "bootstrap deletion and everything after".

I'm not against the division of the install process into smaller functions or separate files.  My inclination here is to not add new phases and to open an issue to condense the phases that we do have.  WDYT?

jim-minter replied:
> I wanted to have more steps so it's easier to understand where we failed during installation without looking at logs. It might also be useful if we decide to gather stats to make decisions on areas where we need to improve, etc

I agree with this, I think we should aim to achieve it outside of the phase mechanism.  Could I ask you to open two issues - one to get the number of phases down, the second to implement something to address ^

m1kola replied:
Created https://github.com/Azure/ARO-RP/issues/106 and https://github.com/Azure/ARO-RP/issues/107

m1kola commented on pkg/install/2-removebootstrap.go:
```diff
@@ -63,6 +66,15 @@ func (i *Installer) removeBootstrap(ctx context.Context) error {
+			return err
+		}
+
+		i.updateConsoleBranding(ctx, cli)
```
`updateConsoleBranding` accepts a clientset interface, so should be easier to mock. But we will need to reconsider several of bits in the package, once we start addming tests.

jim-minter commented:
nice work!
